x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -316,13 +317,18 @@ public void authorize(
                     parentAuthorization
                 );
                 final AuthorizationEngine engine = getAuthorizationEngine(authentication);
-                final ActionListener<AuthorizationInfo> authzInfoListener = wrapPreservingContext(
-                    listener.delegateFailureAndWrap((l, authorizationInfo) -> {
-                        threadContext.putTransient(AUTHORIZATION_INFO_KEY, authorizationInfo);
-                        maybeAuthorizeRunAs(requestInfo, auditId, authorizationInfo, l);
-                    }),
-                    threadContext
-                );
+                final ActionListener<AuthorizationInfo> authzInfoListener = wrapPreservingContext(ActionListener.wrap(authorizationInfo -> {
+                    threadContext.putTransient(AUTHORIZATION_INFO_KEY, authorizationInfo);
+                    maybeAuthorizeRunAs(requestInfo, auditId, authorizationInfo, listener);
+                }, e -> {
+                    if (e instanceof ElasticsearchAccessRestrictedException) {
+                        auditTrailService.get()
```

Comment:
Based on our discussion (and also for completeness), I've added access denied audit log. We can re-evaluate this decision and most probably introduce dedicated audit even in a follow-up work when we move the role resolution earlier in the authorization flow.